projects/grpc-client-server/prober_server/main.go

```diff
@@ -22,26 +22,64 @@ type server struct {
 	pb.UnimplementedProberServer
 }
 
-func (s *server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
-	// TODO: support a number of repetitions and return average latency
-	start := time.Now()
-	_, _ = http.Get(in.GetEndpoint())	// TODO: add error handling here and check the response code
-	elapsed := time.Since(start)
-	elapsedMsecs := float32(elapsed / time.Millisecond)
-
-	return &pb.ProbeReply{LatencyMsecs: elapsedMsecs}, nil
+func (s *server) DoProbes(ctx context.Context, request *pb.ProbeRequest) (*pb.ProbeReply, error) {
+	// get the number of probes from the probe request
+	numberOfProbes := request.GetNumberOfProbes()
+	// fmt.Printf("DoProbes | numberOfProbes %v\n", numberOfProbes)
+
+	// initialise a total time
+	var totalTime time.Duration
+
+	// support a number of repetitions and return average latency
+	for i := 0; i < int(numberOfProbes); i++ {
+		startTime := time.Now()
+		// fmt.Printf("DoProbes | startTime %v\n", startTime)
+
+		// make the request to the endpoint
+		response, err := http.Get(request.GetEndpoint())
+
+		// if the request errors
+		if err != nil {
+			return nil, fmt.Errorf("error: DoProbes request failed: %v", err)
+		}
+
+		// remember to close the response body
+		defer response.Body.Close()
```

Comment:
This is also an interesting question! You're currently not including the time to download the response body in the probe time. Is it useful to include that time? Is it not? When might each be useful?

If you're _not_ going to read the body, you may want to make a HEAD request rather than a GET, to avoid the server needing to send it.

Reply:
yes, HEAD is a much better idea, this is a new concept to me (not on the main cyf sdc)

it is just a latency test, so the actual body is not important(?)

also say the GET request wanted to return a body with 100mb of data... would that affect the latency? I am assuming yes(?)

does the GET request only "complete" when the full body has been sent down

I will look into this

Reply:
> also say the GET request wanted to return a body with 100mb of data... would that affect the latency?

That depends on what latency we're interested in! Maybe we're interested in "How long does it take for me to create a connection" (i.e. round-trip-time / ping time - known as "time to first byte" or TTFB). Maybe we're interested in "How long does it take for me to download a file" (which will factor in not just the distance to a server, but also the bandwidth available to it - known as "time to last byte" or TTLB).

So it kind of depends on why we're testing this. Are we making sure something is up? Are we trying to provide an (estimated) upper bound on actually using a service?